sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -86,11 +86,31 @@ package object expressions  {
   /**
    * Helper functions for working with `Seq[Attribute]`.
    */
-  implicit class AttributeSeq(attrs: Seq[Attribute]) {
+  implicit class AttributeSeq(val attrs: Seq[Attribute]) {
     /** Creates a StructType with a schema matching this `Seq[Attribute]`. */
     def toStructType: StructType = {
       StructType(attrs.map(a => StructField(a.name, a.dataType, a.nullable)))
     }
+
+    private lazy val inputArr = attrs.toArray
+
+    private lazy val inputToOrdinal = {
+      val map = new java.util.HashMap[ExprId, Int](inputArr.length * 2)
```

Comment:
Why ```*2``` is necessary?
I think that the size of map's entry is up to ```attrs.size``` since the max number of calling ```map.put()``` is equal to ``attrs.size```. Is ```attrs.size``` equal to ```inputArr.legnth```?